### PR TITLE
Openssl 1.0.1 domain name matching tweaks

### PR DESCRIPTION
**Pull request by iximeow**
This is two distinct changes:
* [`76d996c`](https://github.com/sfackler/rust-openssl/pull/1378/commits/76d996c380334217efba23ceddc050c045ababbb) makes hostname comparison case-insensitive. Any poor soul using openssl 1.0.1 may like to see this. It also adds some checks to show what hostnames do/don't match by `matches_dns`.
* [`159922b`](https://github.com/sfackler/rust-openssl/pull/1378/commits/159922b514979aa437060e906981da9bea72268b) then removes partial wildcard support from the openssl 1.0.1 custom verify routine. This disagrees with the original libcurl implementation but is consistent with openssl 1.0.2+ behavior.

I'd recommend reading the commits one after the other, particularly because if there's any hesitation about partial wildcards the first commit should be fine on its own.

edit2: oh, clippy did not like this. I didn't touch openssl-sys and I don't see recent changes to `openssl-sys/build/main.rs` so I'm not sure why clippy chose me to be mad at? :(

### CODE REVIEW

**Comment by sfackler** [comment]
I'll fix the clippy bits - we use the current stable compiler so there tend to be some extra warnings when new stable releases happen.

File: openssl/src/ssl/connector.rs [hunk 1]
```diff
@@ -552,12 +552,19 @@ cfg_if! {
                 let hostname_label = &hostname[..hostname_label_end];
 
                 // check the prefix of the first label
-                if !hostname_label.starts_with(wildcard_prefix) {
+                if !hostname_label[..wildcard_location].eq_ignore_ascii_case(wildcard_prefix) {
```
Comment (sfackler) [inline openssl/src/ssl/connector.rs]: Do we know at this point that hostname_label is at least as long as wildcard_location?

File: openssl/src/ssl/connector.rs [hunk 2]
```diff
@@ -502,11 +502,6 @@ cfg_if! {
             }
 
             fn matches_wildcard(pattern: &str, hostname: &str) -> Option<bool> {
-                // internationalized domains can't involved in wildcards
```
Comment (sfackler) [inline openssl/src/ssl/connector.rs:505]: Why is this check removed?

Comment (sfackler) [reply openssl/src/ssl/connector.rs:505]: Oh right, because the xn-- prefix implies that a wildcard has to be partial.

**Comment by sfackler** [comment]
The clippy error should be fixed if you rebase.

File: openssl/src/ssl/connector.rs (hunk 1 above)
Comment (iximeow) [reply openssl/src/ssl/connector.rs]: indeed, we do not. Good catch. There's [a check for that case](https://github.com/sfackler/rust-openssl/pull/1378/commits/76d996c380334217efba23ceddc050c045ababbb#diff-1b6ced8dfb91d53c0d3c93878fca256ae1431551161033aa3a264992e018829bR555-R559) and [more tests](https://github.com/sfackler/rust-openssl/pull/1378/commits/76d996c380334217efba23ceddc050c045ababbb#diff-1b6ced8dfb91d53c0d3c93878fca256ae1431551161033aa3a264992e018829bR601-R605) for partial wildcards now.

**Comment by iximeow** [comment]
Done, and CI agrees with these changes now \o/

**Comment by sfackler** [comment]
Thanks!

**Pull request merged**